src/main/java/com/aws/greengrass/lifecyclemanager/KernelLifecycle.java

```diff
@@ -457,11 +457,24 @@ public void startupAllServices() {
      * Shutdown all services in dependency order.
      *
      * @param timeoutSeconds timeout seconds for waiting all services to shutdown. Use -1 to wait infinitely.
+     * @param serviceNames Configuration of services yet to be merged
      */
     @SuppressWarnings("PMD.AvoidCatchingThrowable")
-    public void stopAllServices(int timeoutSeconds) {
-        GreengrassService[] d = kernel.orderedDependencies().toArray(new GreengrassService[0]);
+    public void stopAllServices(int timeoutSeconds, Set<String> serviceNames) {
+        HashSet<String> runningServices = kernel.orderedDependencies().stream().map(GreengrassService::getName)
+                .collect(Collectors.toCollection(HashSet::new));
+        ArrayList<GreengrassService> newServices = serviceNames.stream()
+                .filter(service -> !runningServices.contains(service))
+                .collect(Collectors.toCollection(ArrayList::new))
+                .stream()
+                .map(kernel::locateIgnoreError)
```

Comment:
You should be able to map then filter and not need to collect and recreate another stream in between. `locateIgnoreError` will simply return the running instance of a service if the service exists IIRC.

Reply:
Also, why does this only need to be done for new services and not running services, could running services not have a lifecycle script change in the config merge that could trigger its lifecycle listener, why won't we need to wait for that?

Reply:
I did not investigate this, will do that now

Reply:
I have the answer to my second question, since the running services are in the orderedDependencies list, nucleus shutdown will shut them down and wait as it is.

Reply:
I was thinking something else, when `newConfig` is merged after `softShutdown`, if an existing service has one of its lifecycle steps modified then that should trigger a `dependencyInit` for the lifecycle listener. This is not something that I investigated.

Reply:
Oh but the service is closed

Reply:
Maybe what you're saying is possible too, we should check whether that does or doesn't happen